packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useRef,
+	useMemo,
+	useCallback,
+	useEffect,
+	useReducer,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+const useIsomorphicLayoutEffect =
+	typeof window !== 'undefined' ? useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+/**
+ * Custom react hook for retrieving props from registered selectors.
+ *
+ * In general, this custom React hook follows the
+ * [rules of hooks](https://reactjs.org/docs/hooks-rules.html).
+ *
+ * @param {Function} _mapSelect  Function called on every state change. The
```

Comment:
Minor: Considering that this is enshrined in the [public-facing documentation](https://github.com/WordPress/gutenberg/blob/master/packages/data/README.md#useSelect), I think we could have optimized for this to be the `mapSelect`, either choosing `_mapSelect` or (preferably, if one exists) a better name for the internal reference.

I guess it depends on your preference for or against "modifying" the argument, but since the arguments aren't `const`, you could always re-define:

```js
export default function useSelect( mapSelect, deps ) {
	mapSelect = useCallback( mapSelect, deps );
```

Reply:
hmm true.